specs/ExecuteScriptWithResult.md

```diff
@@ -0,0 +1,315 @@
+get the script execution result as a JSON string or as a string value if execution succeeds, and can be used to get the exception when
+execution failed.
+
+# Examples
+The following code snippets demonstrate how the ExecuteScriptWithResult can be used:
+## Win32 C++
+``` cpp
+// Tools function to generate the script code
+// Using std::wstringstream to generate script code,
+// it will generate the code like
+// 'let str_0 = "This is a string"; let n_0= str_0.replace(/a/i, "microsoft"); n_0;'
+static std::wstring GenerateScriptCode(LPCWSTR str, LPCWSTR reg, LPCWSTR item)
+{
+    if (str == nullptr || reg == nullptr || item == nullptr)
+    {
+        return L"";
+    }
+
+    // This variable is used to ensure that the
+    // variables in the script are unique.
+    static int idx = 0;
+
+    static std::wstringstream sw;
+    sw.clear();
+
+    sw << L"let str_" << idx << L" = \"" << str << L"\"; let n_" << idx << L"= str_" << idx
+       << L".replace(" << reg << L", \"" << item << L"\"); n_" << idx << L";";
+    ++idx;
+    return sw.str();
+}
+
+// This is a demo that uses regular expressions in 
+// javascript to complete string replacement, it will handle
+// the case of successful execution and execution exception
+void MatchRegWithScript(wil::com_ptr<ICoreWebView2> webView
+    , LPCWSTR str
+    , LPCWSTR reg
+    , LPCWSTR item)
+{
+    wil::com_ptr<ICoreWebView2_10> webview2 = webView.try_query<ICoreWebView2_10>();
+
+    auto scriptCode = GenerateScriptCode(str, reg, item);
+    webview2->ExecuteScriptWithResult(
+        scriptCode.c_str(),
+        Callback<ICoreWebView2ExecuteScriptWithResultCompletedHandler>(
+            [](
+                HRESULT errorCode, ICoreWebView2ExecuteScriptResult* result) -> HRESULT
+            {
+                if (errorCode != S_OK || result == nullptr)
+                {
+                    MessageBox(
+                        nullptr, L"Get execute status failed!", L"ExecuteScript Result", MB_OK);
+                }
+                else
+                {
+                    wil::unique_cotaskmem_string stringData;
+
+                    BOOL isSuccess;
+                    result->get_IsSuccess(&isSuccess);
+                    // Here is the successful execution.
+                    // We will use a MessageBox to print the replaced result.
+                    if (isSuccess)
+                    {
+                        // We try to use TryGetResultAsString to get the string result here.
+                        // Since the javascript platform's `string.replace` returns a string,
+                        // the call here will succeed.
+                        // If the script is replaced by `string.search`, the function will
+                        // return an int and the call will fail here.
+                        if (result->TryGetResultAsString(&stringData) != S_OK)
```

Comment:
Some things suggest this doesn't fit the Try pattern.  Win32 doesn't use a "succeeded" return value indicating usefulness of the out-param.  .NET version looks like it's returning a possibly-null reference (which could be just GetResultAsString()).
Should we drop the Try prefix here?  Otherwise update the signature so it doesn't fail when there's no string.

Reply:
Yes good point. This is copying what we had from WebMessage event args which seems like may be wrong.

The result could be a string or some other JavaScript type so a Try- method seems appropriate rather than a ResultAsString property that can throw? Since winrt has no difference between empty string and null, and an empty string is a valid string that could be returned, seems like we'd need to return a Boolean not a String

`Boolean TryGetResultAsString(String* resultAsString); // Returns false when result is not a string`
`HRESULT TryGetResultAsString([out] LPWSTR* resultAsString, [out, retval] BOOL* resultIsString);`

Or

`String ResultAsString { get; } // Throws when result is not a string`
`[propget] HRESULT ResultAsString([out, retval] LPWSTR* value); // Returns E_INVALIDARG when result is not a string`

Reply:
I think the `Try-` here is right. The design guidance says that you _don't_ need `Try-` if the value has a reasonable "nullness" value, which `HSTRING` (and thus `winrt::hstring`) doesn't.

As an alternative, maybe this should return `Object` - like `result->get_Result(&insp)` and then have the result be-a `PropertyValue`? Then you could do type-coercion on the various `IPropertyValue::Get...` methods.

Reply:
`Try` makes sense to me, and is important even if string does have a sentinel value, because it's not at all obvious that this could fail. (I would have assumed otherwise that this was just doing to convert the json to a string.)

Reply:
Oh, Sorry I didn't write it clearly in the interface description. The current interface can already distinguish whether the acquisition is successful by the return value. The return value is explained as follows, and I will update it in the interface description.

1. S_OK:          Execution succeed.
2. E_INVALIDARG:  When the script result is not a string type.
3. E_POINTER:     When the `stringResult` is nullptr.

Reply:
Oh, Sorry I didn't write it clearly in the interface description. The current interface can already distinguish whether the acquisition is successful by the return value. The return value is explained as follows, and I will update it in the interface description.

1. S_OK:          Execution succeed.
2. E_INVALIDARG:  When the execution result obtained by using the `get_Succeeded()` interface is successful.
3. E_POINTER:     When the `exception` is nullptr.

Reply:
Does this mean the TryGetResultAsString method is returning a Failure (E_* HRESULT) if there is no string to return?  That doesn't align with my expectations of calling a method that will "Try" to get the string for me.  It should be able to return "there is no string" on a success path.

Reply:
OK, that make sence, I'll change it as follow.
```
HRESULT TryGetStringResult([out] LPWSTR* result, [out, retval] BOOL* value);
```